cachi2/core/package_managers/gomod.py

```diff
@@ -79,10 +79,146 @@ class GoPackage(_GolangModel):
 
     import_path: str
     standard: bool = False
-    module: Optional[GoModule]
+    module: Optional[ParsedModule]
     deps: list[str] = []
 
 
+class Module(NamedTuple):
+    """A Go module with relevant data for the SBOM generation"""
+    name: str
+    # name of the module before it was replaced
+    original_name: str
+    version: str
+    # The real URL to reach the module, differs from name in case of local replacements
+    url: str
+    main: bool = False
+
+    @property
+    def purl(self) -> str:
+        return f"pkg:golang/{self.url}@{self.version}?type=module"
+
+    def to_component(self) -> Component:
+        return Component(name=self.name, version=self.version, purl=self.purl)
+
+
+class Package(NamedTuple):
+    """A Go package with relevant data for the SBOM generation"""
+    name: str
+    module: Optional[Module]
+    standard: bool = False
+
+    @property
+    def relative_path(self) -> str:
+        return self.name.replace(self.module.name, "")
+
+    @property
+    def purl(self) -> str:
+        if self.standard:
+            return f"pkg:golang/{self.name}"
+
+        return f"pkg:golang/{self.module.url}{self.relative_path}@{self.module.version}?type=package"
+
+    def to_component(self) -> Component:
+        if self.standard:
+            version = None
+        else:
+            version = self.module.version
+
+        return Component(name=self.name, version=version, purl=self.purl)
+
+
+class ModuleFactory(NamedTuple):
```

Comment:
I went a little Java-ish here :coffee: :blush: 

I was bothered that the `from_parsed_module` function would always need to receive the main_module and app_dir as params, and I also liked the nesting of auxiliary functions such as `_resolve_url_for_local_replacement`, as it gives a better sense of scope.